### DISCUSSION TITLE
Default member type (number, string, or symbol?)

### DISCUSSION THREAD

**Discussion by Jack-Works** [Jack-Works/proposal-enum#16] [General]
What should be the default member type of an enum?

We may specify the type by a flow.js-like syntax:

```js
enum A for symbol { A, B }
enum B for string { A, B }
enum C for number { A = 1, B = 2 }
enum D for bigint { A = 1n, B = 2n }
```

> Note: `enum Identifier of _type_` is used by flow.js. To avoid syntax collision, I choose `for` as the keyword.

But what should happen if no member type is specified?

```js
enum A { A, B }
```

Should it be number, string, symbol, or bigint?

If we enforce developers to specify a member type, we can ignore this problem and it also creates a different syntax to resolve the syntax collision with TypeScript and flow.js (#14)

**Comment by rbuckton** [comment]
Despite it being in conflict with TypeScript enums, I believe that there should either be no default. From TypeScript's perspective, that would be the easiest way to distinguish between a TypeScript `enum` in an existing declaration file vs a native `enum`. 

If a default is preferred, in my opinion the next best option would for it to be an ADT enum. In that case, TypeScript would:

1. Always assume an `enum` without a `for` clause is `for number` for backwards compatibility with existing type definitions on DefinitelyTyped.
2. Always emit an `enum` with a `for` clause to a definition file, even if one isn't present on the `enum` declaration the user wrote.

**Comment by Jack-Works** [comment]
I support enforcing the type to disambiguate the syntax.

Enforcing developers to specify the default value type will have another problem. Because TypeScript and Flow.js can be used without types specified, it will encourage developers to use TS/flow enum if it is possible (because it is shorter).

**Reply by ljharb** [reply]
How does that affect mixed types?

**Reply by Jack-Works** [reply]
Yeah, that's also another problem. If we enforce a type, how it will work with mixed types? I don't think `enum X of mixed` is a good design.

**Reply by ljharb** [reply]
That suggests that type enforcement itself isn't a good design; that's something linters can cover.

**Reply by Jack-Works** [reply]
So let's only use `enum T of symbols` (things like that) for auto-generated enum values? Then what should be the default member type (or is that question even meaningful? e.g. we explicitly require an "autoEnum" generator or initialization field, like one of `enum T via MyEnumFactory('symbol')` or `enum T { A = 1 }`)?

**Reply by ljharb** [reply]
I’m not even sure autogenerating values is a good idea, tbh

**Reply by Jack-Works** [reply]
> I’m not even sure autogenerating values is a good idea, tbh

At least for symbols and strings?
